File: examples/javascript/limit-testing/limits-test.ts
import { task } from "@capsule-run/sdk";

/**
 * This task should TIMEOUT because it sleeps longer than the configured timeout.
 */
export const testTimeout = task(
  {
    name: "testTimeout",
    timeout: "2s",
    compute: "HIGH"
  },
  async (): Promise<string> => {
    await new Promise(resolve => setTimeout(resolve, 10000));

    console.log("Sleep completed (this should NOT print)");
    return "TIMEOUT TEST FAILED - Should have timed out!";
  }
);

/**
 * This task should EXCEED compute limits with intensive CPU operations.
 */
export const testComputeLimitLow = task(
  {
    name: "testComputeLimitLow",
    compute: "LOW",
    timeout: "60s"
  },
  (): string => {
    let result = 0;

    for (let i = 0; i < 100000000; i++) {
      result += Math.sqrt(i) * Math.sin(i) * Math.cos(i);

      if (i % 1000 === 0) {
        result = result % 1000000;
      }
    }

    console.log("Computation completed (this should NOT print)");
    return `COMPUTE TEST FAILED - Should have run out of fuel! Result: ${result}`;
  }
);

/**
 * This task should EXCEED MEDIUM compute limits.
 */
export const testComputeLimitMedium = task(
  {
    name: "testComputeLimitMedium",
    compute: "MEDIUM",
    timeout: "60s"
  },
  (): string => {
    let result = 0;

    for (let i = 0; i < 10000000000; i++) {
      result += Math.sqrt(i) * Math.sin(i) * Math.cos(i) * Math.tan(i);

      if (i % 10000 === 0) {
        result = result % 1000000;
      }
    }

    console.log("Computation completed (this should NOT print)");
    return `COMPUTE TEST FAILED - Should have run out of fuel! Result: ${result}`;
  }
);

/**
 * This task should EXCEED RAM limits by allocating large arrays.
 */
export const testRAMLimit = task(
  {
    name: "testRAMLimit",
    ram: "10MB",
    compute: "HIGH",
    timeout: "30s"
  },
  (): string => {
    const arrays: number[][] = [];

    try {
      for (let i = 0; i < 1000; i++) {
        const largeArray = new Array(1000000).fill(0).map((_, idx) => idx * Math.random());
        arrays.push(largeArray);

        if (i % 10 === 0) {
          console.log(`📊 Allocated ${i} arrays (~${i * 8}MB)`);
        }
      }

      console.log("Memory allocation completed (this should NOT print)");
      return `RAM TEST FAILED - Should have run out of memory! Arrays: ${arrays.length}`;
    } catch (error) {
      console.log(`JavaScript error: ${error}`);
      throw error;
    }
  }
);

/**
 * This task tests multiple limits simultaneously:
 * - Very low timeout
 * - Low compute
 * - Low RAM
 */
export const testCombinedLimits = task(
  {
    name: "testCombinedLimits",
    timeout: "3s",
    compute: "LOW",
    ram: "5MB"
  },
  async (): Promise<string> => {
    const data: number[][] = [];
    let computeResult = 0;

    for (let i = 0; i < 100; i++) {
      data.push(new Array(100000).fill(0).map(() => Math.random()));

      for (let j = 0; j < 1000000; j++) {
        computeResult += Math.sqrt(j) * Math.sin(j);
      }

      await new Promise(resolve => setTimeout(resolve, 100));
    }

    console.log("All operations completed (this should NOT print)");
    return `COMBINED TEST FAILED - Should have hit a limit! Data: ${data.length}, Compute: ${computeResult}`;
  }
);

/**
 * Tests timeout with multiple async operations.
 */
export const testAsyncTimeout = task(
  {
    name: "testAsyncTimeout",
    timeout: "5s",
    compute: "MEDIUM"
  },
  async (): Promise<string> => {
    const delays = [2000, 2000, 2000, 2000, 2000];

    for (let i = 0; i < delays.length; i++) {
      await new Promise(resolve => setTimeout(resolve, delays[i]));
      console.log(`✓ Delay ${i + 1} completed`);
    }

    console.log("All delays completed (this should NOT print)");
    return "ASYNC TIMEOUT TEST FAILED - Should have timed out!";
  }
);

export const testImmediateFuelExhaustion = task(
  {
    name: "testImmediateFuelExhaustion",
    compute: "LOW",
    timeout: "60s"
  },
  (): string => {
    let result = 0;
    const iterations = 1000000000;

    for (let i = 0; i < iterations; i++) {
      result += Math.pow(Math.sqrt(i), 2) * Math.sin(i) * Math.cos(i) * Math.tan(i) * Math.log(i + 1);

      if (i % 100000000 === 0 && i > 0) {
        console.log(`  Progress: ${(i / iterations * 100).toFixed(1)}%`);
      }
    }

    console.log("Completed (this should NOT print)");
    return `FUEL TEST FAILED - Result: ${result}`;
  }
);

/**
 * Main function to demonstrate all limit tests.
 *
 * Uncomment individual tests to run them one at a time.
 * Each test is designed to FAIL with a specific limit error.
 */
export const main = task(
  {
    name: "main",
    compute: "HIGH",
    timeout: "120s"
  },
  async (): Promise<void> => {
    // Test 1: Timeout
    console.log("\n🧪 Running Test 1: Timeout Limit");
    await testTimeout();

    // Test 2: Low Compute
    console.log("\n🧪 Running Test 2: LOW Compute Limit");
    testComputeLimitLow();

    // Test 3: Medium Compute
    console.log("\n🧪 Running Test 3: MEDIUM Compute Limit");
    testComputeLimitMedium();

    // Test 4: RAM Limit
    console.log("\n🧪 Running Test 4: RAM Limit");
    testRAMLimit();

    // Test 5: Combined Limits
    console.log("\n🧪 Running Test 5: Combined Limits");
    await testCombinedLimits();

    // Test 6: Async Timeout
    console.log("\n🧪 Running Test 6: Async Timeout");
    await testAsyncTimeout();

    // Test 7: Immediate Fuel Exhaustion
    console.log("\n🧪 Running Test 7: Immediate Fuel Exhaustion");
    testImmediateFuelExhaustion();

    console.log("\n limit testing completed.");
  }
);

File: examples/python/limit-testing/limits_test.py
import math
import time
from capsule import task


@task(name="test_timeout", timeout="2s", compute="HIGH")
def test_timeout() -> str:
    """
    This task should TIMEOUT because it sleeps longer than the configured timeout.
    """
    time.sleep(10)

    print("Sleep completed (this should NOT print)")
    return "TIMEOUT TEST FAILED - Should have timed out!"


@task(name="test_compute_limit_low", compute="LOW", timeout="60s")
def test_compute_limit_low() -> str:
    """
    This task should EXCEED compute limits with intensive CPU operations.
    """
    result = 0.0

    for i in range(100_000_000):
        result += math.sqrt(i) * math.sin(i) * math.cos(i)

        if i % 1000 == 0:
            result = result % 1_000_000

    print("Computation completed (this should NOT print)")
    return f"COMPUTE TEST FAILED - Should have run out of fuel! Result: {result}"


@task(name="test_compute_limit_medium", compute="MEDIUM", timeout="60s")
def test_compute_limit_medium() -> str:
    """
    This task should EXCEED MEDIUM compute limits.
    """
    result = 0.0

    for i in range(10_000_000_000):
        result += math.sqrt(i) * math.sin(i) * math.cos(i) * math.tan(i)

        if i % 10000 == 0:
            result = result % 1_000_000

    print("Computation completed (this should NOT print)")
    return f"COMPUTE TEST FAILED - Should have run out of fuel! Result: {result}"


@task(name="test_ram_limit", ram="10MB", compute="HIGH", timeout="30s")
def test_ram_limit() -> str:
    """
    This task should EXCEED RAM limits by allocating large lists.
    """
    arrays = []

    try:
        for i in range(1000):
            large_array = [idx * 0.5 for idx in range(1_000_000)]
            arrays.append(large_array)

            if i % 10 == 0:
                print(f"📊 Allocated {i} arrays (~{i * 8}MB)")

        print("Memory allocation completed (this should NOT print)")
        return f"RAM TEST FAILED - Should have run out of memory! Arrays: {len(arrays)}"
    except Exception as error:
        print(f"Python error: {error}")
        raise error


@task(name="test_combined_limits", timeout="3s", compute="LOW", ram="5MB")
def test_combined_limits() -> str:
    """
    This task tests multiple limits simultaneously:
    - Very low timeout
    - Low compute
    - Low RAM
    """
    data = []
    compute_result = 0.0

    for i in range(100):
        data.append([j * 0.5 for j in range(100_000)])

        for j in range(1_000_000):
            compute_result += math.sqrt(j) * math.sin(j)

        time.sleep(0.1)

    print("All operations completed (this should NOT print)")
    return f"COMBINED TEST FAILED - Should have hit a limit! Data: {len(data)}, Compute: {compute_result}"


@task(name="test_async_timeout", timeout="5s", compute="MEDIUM")
def test_async_timeout() -> str:
    """
    Tests timeout with multiple sleep operations.
    """
    delays = [2.0, 2.0, 2.0, 2.0, 2.0]

    for i, delay in enumerate(delays):
        time.sleep(delay)

    print("All delays completed (this should NOT print)")
    return "ASYNC TIMEOUT TEST FAILED - Should have timed out!"


@task(name="test_immediate_fuel_exhaustion", compute="LOW", timeout="60s")
def test_immediate_fuel_exhaustion() -> str:
    """
    This task should run out of fuel almost immediately.
    """
    result = 0.0
    iterations = 1_000_000_000

    for i in range(iterations):
        result += (
            math.pow(math.sqrt(i), 2)
            * math.sin(i)
            * math.cos(i)
            * math.tan(i)
            * math.log(i + 1)
        )

    print("Completed (this should NOT print)")
    return f"FUEL TEST FAILED - Result: {result}"


@task(name="main", compute="HIGH", timeout="120s")
def main() -> None:
    """
    Main function to demonstrate all limit tests.

    Uncomment individual tests to run them one at a time.
    Each test is designed to FAIL with a specific limit error.
    """
    # Test 1: Timeout
    print("\n🧪 Running Test 1: Timeout Limit")
    test_timeout()

    # Test 2: Low Compute
    print("\n🧪 Running Test 2: LOW Compute Limit")
    test_compute_limit_low()

    # Test 3: Medium Compute
    print("\n🧪 Running Test 3: MEDIUM Compute Limit")
    test_compute_limit_medium()

    # Test 4: RAM Limit
    print("\n🧪 Running Test 4: RAM Limit")
    test_ram_limit()

    # Test 5: Combined Limits
    print("\n🧪 Running Test 5: Combined Limits")
    test_combined_limits()

    # Test 6: Async Timeout
    print("\n🧪 Running Test 6: Async Timeout")
    test_async_timeout()

    # Test 7: Immediate Fuel Exhaustion
    print("\n🧪 Running Test 7: Immediate Fuel Exhaustion")
    test_immediate_fuel_exhaustion()

    print("\n✅ Limit testing completed.")
